R/check_smooth.R

```diff
@@ -0,0 +1,243 @@
+    if (ismat) attr(x, "matrix") <- TRUE
+    x
+  }
+
+
+  ## `data' and `knots' contain the data needed to evaluate the `terms', `by'
+  ## and `knots' elements of `object'. This routine does so, and returns
+  ## a list with element `data' containing just the evaluated `terms',
+  ## with the by variable as the last column. If the `terms' evaluate matrices,
+  ## then a check is made of whether repeat evaluations are being made,
+  ## and if so only the unique evaluation points are returned in data, along
+  ## with the `index' attribute required to re-assemble the full dataset.
+  knt <- dat <- list()
+  ## should data be processed as for summation convention with matrix arguments?
+  vecMat <- if (is.null(object$xt$sumConv)) TRUE else object$xt$sumConv
+  for (i in 1:length(object$term)) {
+    dat[[object$term[i]]] <- get.var(object$term[i], data, vecMat = vecMat)
+    knt[[object$term[i]]] <- get.var(object$term[i], knots, vecMat = vecMat)
+  }
+  names(dat) <- object$term
+  m <- length(object$term)
+  if (!is.null(attr(dat[[1]], "matrix")) && vecMat) { ## strip down to unique covariate combinations
+    n <- length(dat[[1]])
+    X <- matrix(unlist(dat), n, m)
+    if (is.numeric(X)) {
+      X <- mgcv::uniquecombs(X)
+      if (nrow(X) < n * .9) { ## worth the hassle
+        for (i in 1:m) dat[[i]] <- X[, i] ## return only unique rows
+        attr(dat, "index") <- attr(X, "index") ## index[i] is row of dat[[i]] containing original row i
+      }
+    } ## end if(is.numeric(X))
+  }
+  if (object$by != "NA") {
+    by <- get.var(object$by, data)
+    if (!is.null(by)) {
+      dat[[m + 1]] <- by
+      names(dat)[m + 1] <- object$by
+    }
+  }
+  list(data = dat, knots = knt)
+}
+
+
+
+# https://github.com/cran/mgcv/blob/2db5036f529dff6ec8a4a4ba0d2df1804a4d2668/R/sparse.r#L129
+#' @keywords internal
+mgcv_nearest <- function(k, X, gt.zero = FALSE, get.a = FALSE) {
+
+  insight::check_if_installed("mgcv")
+
+  ## The rows of X contain coordinates of points.
+  ## For each point, this routine finds its k nearest
+  ## neighbours, returning a list of 2, n by k matrices:
+  ## ni - ith row indexes the rows of X containing
+  ##      the k nearest neighbours of X[i,]
+  ## dist - ith row is the distances to the k nearest
+  ##        neighbours.
+  ## a - area associated with each point, if get.a is TRUE
+  ## ties are broken arbitrarily.
+  ## gt.zero indicates that neighbours must have distances greater
+  ## than zero...
+
+  if (gt.zero) {
+    Xu <- mgcv::uniquecombs(X)
+    ind <- attr(Xu, "index") ## Xu[ind,] == X
+  } else {
+    Xu <- X
+    ind <- 1:nrow(X)
+  }
+  if (k > nrow(Xu)) stop("not enough unique values to find k nearest")
+  nobs <- length(ind)
+  n <- nrow(Xu)
+  d <- ncol(Xu)
+  dist <- matrix(0, n, k)
+  if (get.a) a <- 1:n else a <- 1
+
+  # TODO: how to get that without the call to an internal????
+  oo <- .C(mgcv:::C_k_nn,
```

Comment:
The biggest issue is there: I suspect it is an C object (but can't find it in the package https://github.com/cran/mgcv !), and I wonder whether there's a way to bypass that... the problem is that its function, `nearest()`, is not export from mgcv